### PR TITLE
Refine Tile Processing and Expand Tile Configuration

### PR DESCRIPTION
**Pull request by eugene123tw**
### Summary

* Introduce Tile Sampling Feature
* Implement Tile Adaptor
* Integrate OTX Tile Transform: Unfolding Tiles with Customizable Tile Size
* Filtering for Empty Val set Tiles
* Expand Tile Configuration Options
* Incorporate NMS in Tile Postprocessing
* Update Tile-Related Recipes to Accommodate Changes

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/callbacks/tile_sampling_hook.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Tile Sampling Hook for OTXTileTrainDataset."""
+from lightning import Callback, LightningModule, Trainer
+
+
+class TileSamplingHook(Callback):
```
Comment (sungmanc) [inline src/otx/algo/callbacks/tile_sampling_hook.py]: Personally, I prefer to use the word "Callback" for the OTX2.0
```suggestion
class TileSamplingCallback(Callback):
```

Comment (vinnamkim) [inline src/otx/algo/callbacks/tile_sampling_hook.py]: Why do we need this?

File: src/otx/core/data/dataset/tile.py [hunk 2]
```diff
@@ -124,12 +197,16 @@ def get_tiles(self, image: np.ndarray, item: DatasetItem) -> tuple[list[OTXDataE
...
         tile_ds = tile_ds.transform(
-            Tile,
-            grid_size=self.tile_config.grid_size,
+            OTXTileTransform,
+            tile_size=self.tile_config.tile_size,
             overlap=(self.tile_config.overlap, self.tile_config.overlap),
             threshold_drop_ann=0.5,
         )
 
+        if self.dm_subset.name == "val":
```
Comment (vinnamkim) [inline src/otx/core/data/dataset/tile.py:205]: After considering https://github.com/openvinotoolkit/datumaro/pull/1264 and https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/#dataset, I believe that evaluation on no label image is also important for demonstrating model's accuracy on false positives. Therefore, I think that we should not filter an image with empty annotation in validation/test subsets.

@jaegukhyun

File: src/otx/core/data/dataset/tile.py [hunk 3]
```diff
@@ -152,20 +229,30 @@ class OTXTileTrainDataset(OTXTileDataset):
...
-        dataset.dm_subset = dm_subset
-        dataset.ids = [item.id for item in dm_subset]
+        self.dm_subset = DatasetSubset(dm_dataset, dataset.dm_subset.name)
+        self.tile_config = tile_config
+        sampled_subset = self.sample_subset()
+        dataset.dm_subset = sampled_subset
+        dataset.ids = [item.id for item in sampled_subset]
         super().__init__(dataset, tile_config)
 
+    def sample_subset(self) -> DatasetSubset:
```
Comment (vinnamkim) [inline src/otx/core/data/dataset/tile.py]: Why do we need this?

File: src/otx/core/data/dataset/tile.py (hunk 2 above)
Comment (jaegukhyun) [reply src/otx/core/data/dataset/tile.py:205]: I agree on this. OTX1.5 filters empty annotations for train dataset, not for val or test dataset.
And OTX2.0 also only filters for train dataset
https://github.com/openvinotoolkit/training_extensions/blob/c5ad4258a29376b3f9952c9b5441b675dd6f5c8c/src/otx/core/data/pre_filtering.py#L28

Comment (jaegukhyun) [reply src/otx/core/data/dataset/tile.py:205]: I'm not sure whether tiling has same strategy with normal detection training in OTX1.5.
Could you verify this?

File: src/otx/algo/callbacks/tile_sampling_hook.py (hunk 1 above)
Comment (eugene123tw) [reply src/otx/algo/callbacks/tile_sampling_hook.py]: @vinnamkim This PR mainly brings the tiling features from OTX 1.x to here.

Tile sampling is employed to expedite training duration by selecting a portion of samples. Through extensive experimentation, we observed the efficacy (faster training while maintaining good accuracy) of tile sampling particularly for Geti dataset (Vitens). However, numerous tiles can significantly impede training speed. This enhancement primarily targets Geti users who prioritize prompt results. Tiling processes, both in training and inference, are time-consuming. We (Jauguk, Songki, Sungman and I) dedicated 2-3 weeks to optimizing tiling in OTX 1.x. For further insights, you could DM me or refer to the folks involved. I don't think it's ideal to elaborate on all previous activities in this PR.

File: src/otx/core/data/dataset/tile.py (hunk 3 above)
Comment (eugene123tw) [reply src/otx/core/data/dataset/tile.py]: It's used in `TileSamplingHook` which sample tiles from training dataset when epoch starts

File: src/otx/core/data/dataset/tile.py (hunk 2 above)
Comment (eugene123tw) [reply src/otx/core/data/dataset/tile.py:205]: @jaegukhyun We've implemented a filtering mechanism for empty images in both **train** and **val** datasets (excluding the test set). You can find more details on the implementation at this link: [GitHub - OTX Detection Adapters Dataset Module](https://github.com/openvinotoolkit/training_extensions/blob/develop/src/otx/algorithms/detection/adapters/mmdet/datasets/dataset.py#L344). This idea was originally proposed by you in OTX 1.x. Initially, I included all tiles for evaluation, but during tile optimization exercises, we discovered that this approach was too slow. Implementing the filter significantly improved evaluation speed.

@vinnamkim If you believe this filter is unnecessary, I recommend bringing it up with Mark. This feature plays a crucial role in optimizing the speed of tiling. Removing it would likely result in slower tiling processes. Ultimately, the decision regarding whether slower training is acceptable rests with Mark.

Comment (jaegukhyun) [reply src/otx/core/data/dataset/tile.py:205]: I agree on @eugene123tw, since OTX1.5 also filters validation dataset

File: src/otx/algo/callbacks/tile_sampling_hook.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/algo/callbacks/tile_sampling_hook.py]: I still don't know why it is needed. But if so, it seems better to add [RandomSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.RandomSampler) with `num_samples` argument to `torch.DataLoader` at initializing `OTXDataModule` rather than adding this callback. The current way can create a hidden constraint for the configurations (sampling prob < 1.0 should include this callback).

File: src/otx/core/data/dataset/tile.py (hunk 2 above)
Comment (vinnamkim) [reply src/otx/core/data/dataset/tile.py:205]: I'll discuss it with Wonju and Mark

File: src/otx/algo/callbacks/tile_sampling_hook.py (hunk 1 above)
Comment (eugene123tw) [reply src/otx/algo/callbacks/tile_sampling_hook.py]: Thank you for the advice. I've implemented your suggestion by adding `RandomSampler` to train dataloader, which is a better approach. Also, I've removed the Datumaro sampling and tile sampling hook and made other necessary adjustments.

**Review by vinnamkim — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]
